### PR TITLE
EASY-1992 mask IPv6 addresses

### PR DESCRIPTION
**Pull request by jo-pol**
fixes EASY-1992 mask IPv6 addresses

#### When applied it will
* mask IPv6 address and mixed IPv6/IPv4 addresses
* [ ] decide what to mask:
  the full subnet/interface ID's or just a least/most significant portion?
  ICT-Support ticket 2325
* [x] solving merge conflicts might have gone wrong
* 

#### How should this be manually tested?

* Upgrade a module like deposit-api to this version and deploy on `t`easy
* create request(s) the deposit-ui 
* examine the api log

#### related pull requests on github
has merge conflicts with #17

### CODE REVIEW

File: src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala [hunk 1]
```diff
@@ -122,4 +118,62 @@ class MaskerSpec extends FlatSpec with Matchers {
     Masker.formatAuthenticationParameter("other-header" -> Seq("some value")) shouldBe
       "other-header" -> Seq("some value")
   }
+
+  private case class TestCase(address: String, expected: String)
+  private val headCase::tailCases = Seq(
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala]:
```suggestion
  private val headCase::tailCases = List(
```

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/masked/Masker.scala [hunk 2]
```diff
@@ -39,11 +39,18 @@ object Masker {
...
-    remoteAddress.replaceAll("([0-9]+[.]){3}", "**.**.**.")
+    remoteAddress match {
+      case "0:0:0:0:0:0:0:1" | "127.0.0.1" | "::1" | // localhost
+           "0:0:0:0:0:0:0:0" => remoteAddress // unspecified
+      case _ if remoteAddress.contains(".") => remoteAddress
+        .replaceAll("([.][0-9]+){3}$", ".**.**.**")
+      case _ if remoteAddress.matches(":*[0-9A-F]+:+[0-9A-F]+") => remoteAddress
+        .replaceAll("[0-9A-F]+$", "**")
+      case _ => remoteAddress
+        .replaceAll("(:[0-9A-F]+){3}$", ":**:**:**") // e.g: subnet ID + Interface ID of multicast
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/lib/logging/servlet/masked/Masker.scala]: Should we mask the full subnet/interface ID's or just a least/most significant portion?

File: src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala [hunk 3]
```diff
@@ -122,4 +118,62 @@ class MaskerSpec extends FlatSpec with Matchers {
     Masker.formatAuthenticationParameter("other-header" -> Seq("some value")) shouldBe
       "other-header" -> Seq("some value")
   }
+
+  private case class TestCase(address: String, expected: String)
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala]: Have you looked at http://www.scalatest.org/user_guide/table_driven_property_checks. Probably that is exactly what you're trying to achieve here 😉

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/masked/Masker.scala [hunk 4]
```diff
@@ -39,11 +39,18 @@ object Masker {
    * Services without public access might not need to mask.
    */
   def formatRemoteAddress(remoteAddress: String): String = {
-    // TODO https://docs.oracle.com/javase/9/docs/api/java/net/Inet6Address.html
-    // see also:
-    //  * https://stackoverflow.com/questions/3118829/why-request-getremoteaddr-returns-ipv4-or-ipv6-depending-on-context-post-quer
-    //  * https://stackoverflow.com/questions/7589526/what-is-the-ip-format-returned-by-servletrequest-getremoteaddr
-    remoteAddress.replaceAll("([0-9]+[.]){3}", "**.**.**.")
+    remoteAddress match {
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/lib/logging/servlet/masked/Masker.scala:48]: can you identify for each of the cases (except for the first one, of course), what kind of remote addresses you're expecting. A comment like `// IPv4`, `// IPv6`, etc. would do great here!

File: src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala (hunk 3 above)
Comment (jo-pol) [reply src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala]: If only they would fix https://github.com/scalatest/scalatest/issues/178

Comment (jo-pol) [reply src/test/scala/nl/knaw/dans/lib/logging/servlet/masked/MaskerSpec.scala]: Tried version 3.0.5 of scalatest, no improvement on this point.

**Comment by rvanheest** [comment]
@jo-pol and I decided to merge this PR and remove the IPv6 masking later if it turns out not to be necessary after all...

@jo-pol can you fix the merge conflicts?

**Comment by rvanheest** [comment]
Zojuist nogmaals herinnering betreffende de ICT-Support ticket gestuurd. In afwachting van hun reactie.

**Comment by rvanheest** [comment]
@jo-pol there is a merge conflict. Could you resolve that?

**Comment by janvanmansum** [comment]
Closing this one. It has been open since March.

**Pull request closed without merging**